frame/base-fee/src/lib.rs

```diff
@@ -0,0 +1,446 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::{dispatch::DispatchResult, pallet_prelude::*};
+	use frame_system::pallet_prelude::*;
+	use sp_core::U256;
+	use sp_runtime::Permill;
+
+	pub trait BaseFeeThreshold {
+		fn lower() -> Permill;
+		fn upper() -> Permill;
+	}
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event> + IsType<<Self as frame_system::Config>::Event>;
+		/// Lower and upper bounds for increasing / decreasing `BaseFeePerGas`.
+		type Threshold: BaseFeeThreshold;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub base_fee_per_gas: U256,
+		pub is_active: bool,
+		pub elasticity: Permill,
+		_marker: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> GenesisConfig<T> {
+		pub fn new(base_fee_per_gas: U256, is_active: bool, elasticity: Permill) -> Self {
+			Self {
+				base_fee_per_gas,
+				is_active,
+				elasticity,
+				_marker: PhantomData,
+			}
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				// 1 GWEI
+				base_fee_per_gas: U256::from(1_000_000_000),
+				is_active: true,
+				elasticity: Permill::from_parts(125_000),
+				_marker: PhantomData,
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			<BaseFeePerGas<T>>::put(self.base_fee_per_gas);
+			<IsActive<T>>::put(self.is_active);
+		}
+	}
+
+	#[pallet::type_value]
+	pub fn DefaultBaseFeePerGas() -> U256 {
+		U256::from(1_000_000_000)
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn base_fee_per_gas)]
+	pub type BaseFeePerGas<T> = StorageValue<_, U256, ValueQuery, DefaultBaseFeePerGas>;
+
+	#[pallet::type_value]
+	pub fn DefaultIsActive() -> bool {
+		true
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn is_active)]
+	pub type IsActive<T> = StorageValue<_, bool, ValueQuery, DefaultIsActive>;
+
+	#[pallet::type_value]
+	pub fn DefaultElasticity() -> Permill {
+		Permill::from_parts(125_000)
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn elasticity)]
+	pub type Elasticity<T> = StorageValue<_, Permill, ValueQuery, DefaultElasticity>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event {
+		NewBaseFeePerGas(U256),
+		BaseFeeOverflow,
+		IsActive(bool),
+		NewElasticity(Permill),
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(_n: <T as frame_system::Config>::BlockNumber) {
```

Comment:
To keep track of the weight taken by the operations performed in on_finalize, we need to add an empty on_initialize that returns the weight needed by on_finalize.